{
 "cells": [
  {
   "cell_type": "raw",
   "id": "5b1883204b5bb638",
   "metadata": {},
   "source": [
    "!pip install thrift_sasl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pyhive in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (0.7.0)\n",
      "Requirement already satisfied: sqlalchemy in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (2.0.36)\n",
      "Requirement already satisfied: future in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from pyhive) (1.0.0)\n",
      "Requirement already satisfied: python-dateutil in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from pyhive) (2.9.0.post0)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from sqlalchemy) (4.12.2)\n",
      "Requirement already satisfied: six>=1.5 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from python-dateutil->pyhive) (1.16.0)\n",
      "Requirement already satisfied: thrift in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (0.21.0)\n",
      "Requirement already satisfied: six>=1.7.2 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from thrift) (1.16.0)\n",
      "Requirement already satisfied: thrift_sasl in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (0.4.3)\n",
      "Requirement already satisfied: pure-sasl>=0.6.2 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from thrift_sasl) (0.6.2)\n",
      "Requirement already satisfied: six>=1.13.0 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from thrift_sasl) (1.16.0)\n",
      "Requirement already satisfied: thrift>=0.10.0 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from thrift_sasl) (0.21.0)\n",
      "Requirement already satisfied: sqlalchemy in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (2.0.36)\n",
      "Requirement already satisfied: pyhive in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (0.7.0)\n",
      "Requirement already satisfied: thrift in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (0.21.0)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from sqlalchemy) (4.12.2)\n",
      "Requirement already satisfied: future in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from pyhive) (1.0.0)\n",
      "Requirement already satisfied: python-dateutil in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from pyhive) (2.9.0.post0)\n",
      "Requirement already satisfied: six>=1.7.2 in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (from thrift) (1.16.0)\n",
      "Requirement already satisfied: psycopg2-binary in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (2.9.10)\n",
      "Requirement already satisfied: mysql-connector-python in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (9.1.0)\n",
      "Collecting mariadb\n",
      "  Using cached mariadb-1.1.10.tar.gz (84 kB)\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mGetting requirements to build wheel\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[34 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m /bin/sh: mariadb_config: command not found\n",
      "  \u001b[31m   \u001b[0m Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m   File \"/opt/anaconda3/envs/Master/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 353, in <module>\n",
      "  \u001b[31m   \u001b[0m     main()\n",
      "  \u001b[31m   \u001b[0m   File \"/opt/anaconda3/envs/Master/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 335, in main\n",
      "  \u001b[31m   \u001b[0m     json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "  \u001b[31m   \u001b[0m                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/opt/anaconda3/envs/Master/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py\", line 118, in get_requires_for_build_wheel\n",
      "  \u001b[31m   \u001b[0m     return hook(config_settings)\n",
      "  \u001b[31m   \u001b[0m            ^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-build-env-whtpk_4m/overlay/lib/python3.12/site-packages/setuptools/build_meta.py\", line 333, in get_requires_for_build_wheel\n",
      "  \u001b[31m   \u001b[0m     return self._get_build_requires(config_settings, requirements=[])\n",
      "  \u001b[31m   \u001b[0m            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-build-env-whtpk_4m/overlay/lib/python3.12/site-packages/setuptools/build_meta.py\", line 303, in _get_build_requires\n",
      "  \u001b[31m   \u001b[0m     self.run_setup()\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-build-env-whtpk_4m/overlay/lib/python3.12/site-packages/setuptools/build_meta.py\", line 521, in run_setup\n",
      "  \u001b[31m   \u001b[0m     super().run_setup(setup_script=setup_script)\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-build-env-whtpk_4m/overlay/lib/python3.12/site-packages/setuptools/build_meta.py\", line 319, in run_setup\n",
      "  \u001b[31m   \u001b[0m     exec(code, locals())\n",
      "  \u001b[31m   \u001b[0m   File \"<string>\", line 27, in <module>\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-install-33dmx81r/mariadb_3bfd3237e9f34eefb9258f3d2102a5b5/mariadb_posix.py\", line 62, in get_config\n",
      "  \u001b[31m   \u001b[0m     cc_version = mariadb_config(config_prg, \"cc_version\")\n",
      "  \u001b[31m   \u001b[0m                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  \u001b[31m   \u001b[0m   File \"/private/var/folders/b3/kyy095_11m18njbjm9rwdk880000gn/T/pip-install-33dmx81r/mariadb_3bfd3237e9f34eefb9258f3d2102a5b5/mariadb_posix.py\", line 28, in mariadb_config\n",
      "  \u001b[31m   \u001b[0m     raise EnvironmentError(\n",
      "  \u001b[31m   \u001b[0m OSError: mariadb_config not found.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m This error typically indicates that MariaDB Connector/C, a dependency which\n",
      "  \u001b[31m   \u001b[0m must be preinstalled, is not found.\n",
      "  \u001b[31m   \u001b[0m If MariaDB Connector/C is not installed, see installation instructions\n",
      "  \u001b[31m   \u001b[0m If MariaDB Connector/C is installed, either set the environment variable\n",
      "  \u001b[31m   \u001b[0m MARIADB_CONFIG or edit the configuration file 'site.cfg' to set the\n",
      "  \u001b[31m   \u001b[0m  'mariadb_config' option to the file location of the mariadb_config utility.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mGetting requirements to build wheel\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "Requirement already satisfied: pymysql in /opt/anaconda3/envs/Master/lib/python3.12/site-packages (1.1.1)\n",
      "Collecting py4j\n",
      "  Downloading py4j-0.10.9.7-py2.py3-none-any.whl.metadata (1.5 kB)\n",
      "Downloading py4j-0.10.9.7-py2.py3-none-any.whl (200 kB)\n",
      "Installing collected packages: py4j\n",
      "Successfully installed py4j-0.10.9.7\n"
     ]
    }
   ],
   "source": [
    "!pip install pyhive sqlalchemy\n",
    "!pip install thrift\n",
    "!pip install thrift_sasl\n",
    "!pip install sqlalchemy pyhive thrift\n",
    "!pip install psycopg2-binary\n",
    "!pip install mysql-connector-python\n",
    "!pip install mariadb\n",
    "!pip install pymysql\n",
    "!pip install py4j\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "9cfff3d4fb6c7f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "import pandas as pd\n",
    "\n",
    "# Crea la conexión usando SQLAlchemy\n",
    "\n",
    "engine = create_engine('hive://localhost:8084/default')\n",
    "\n",
    "conn_postgre_insercion = create_engine('postgresql://hive:password@localhost:5432/postgres')\n",
    "\n",
    "\n",
    "conn_mariadb_insercion = create_engine('mariadb+pymysql://root:my_password@localhost:3307/mariaGESDB')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "db47f112",
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2\n",
    "conn_postgre = psycopg2.connect(\n",
    "    dbname='postgres',\n",
    "    user='hive',\n",
    "    password='password',\n",
    "    host='localhost',\n",
    "    port='5432'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6d5f8c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn_postgre.autocommit=True\n",
    "cursor = conn_postgre.cursor()\n",
    "query = query = \"\"\" CREATE TABLE viviendas (\n",
    "    propertyCode INT PRIMARY KEY,\n",
    "    url VARCHAR(255),\n",
    "    address VARCHAR(255),\n",
    "    size FLOAT,\n",
    "    floor VARCHAR(50),\n",
    "    province VARCHAR(100),\n",
    "    municipality VARCHAR(100),\n",
    "    district VARCHAR(100),\n",
    "    price FLOAT,\n",
    "    rooms INT,\n",
    "    hasLift BOOLEAN,\n",
    "    hasParking BOOLEAN,\n",
    "    hasTerrace BOOLEAN,\n",
    "    hasSwimmingPool BOOLEAN,\n",
    "    hasAirConditioning BOOLEAN,\n",
    "    hasGarden BOOLEAN,\n",
    "    bathrooms INT\n",
    ");\n",
    " \"\"\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a9162bfd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "560"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "csv = pd.read_csv(r'../data/structured/vivienda.csv')\n",
    "\n",
    "csv.to_sql(name='viviendas', con=conn_postgre_insercion, if_exists='replace', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "92ba3262",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mysql.connector\n",
    "conn_mariadb = mysql.connector.connect(\n",
    "    user=\"root\",\n",
    "    password=\"my_password\",\n",
    "    host=\"localhost\",\n",
    "    database=\"mariaGESDB\",\n",
    "    port=\"3307\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e99753f",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn_mariadb.autocommit=True\n",
    "cursor = conn_mariadb.cursor()\n",
    "query = query = \"\"\" CREATE TABLE location (\n",
    "    propertyCode INT PRIMARY KEY,\n",
    "    latitude FLOAT,\n",
    "    longitude FLOAT\n",
    ");\n",
    " \"\"\"\n",
    "cursor.execute(query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5a5635b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1560"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csv_loc = pd.read_csv(r'../data/structured/location.csv')\n",
    "\n",
    "csv_loc.to_sql(name='location', con=conn_mariadb_insercion, if_exists='replace', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0922a45d",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_create = \"\"\"\n",
    "CREATE EXTERNAL TABLE IF NOT EXISTS viviendas_in_postgres (\n",
    "    propertyCode INT,\n",
    "    url STRING,\n",
    "    address STRING,\n",
    "    size FLOAT,\n",
    "    floor_level STRING,\n",
    "    province STRING,\n",
    "    municipality STRING,\n",
    "    district STRING,\n",
    "    price FLOAT,\n",
    "    rooms INT,\n",
    "    hasLift BOOLEAN,\n",
    "    hasParking BOOLEAN,\n",
    "    hasTerrace BOOLEAN,\n",
    "    hasSwimmingPool BOOLEAN,\n",
    "    hasAirConditioning BOOLEAN,\n",
    "    hasGarden BOOLEAN,\n",
    "    bathrooms INT\n",
    ")\n",
    "STORED BY 'org.apache.hive.storage.jdbc.JdbcStorageHandler'\n",
    "TBLPROPERTIES (\n",
    "    \"hive.sql.database.type\" = \"POSTGRES\",\n",
    "    \"hive.sql.jdbc.driver\" = \"org.postgresql.Driver\",\n",
    "    \"hive.sql.jdbc.url\" = \"jdbc:postgresql://hive4-postgres:5432/postgres\",\n",
    "    \"hive.sql.jdbc.username\" = \"hive\",\n",
    "    \"hive.sql.jdbc.password\" = \"password\",\n",
    "    \"hive.sql.table\" = \"viviendas\"\n",
    ");\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "master",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
